### PR TITLE
[6/n] rustc: transition HIR function bodies from Block to Expr.

### PR DESCRIPTION
**Pull request by eddyb**
_This is part of a series ([prev](https://github.com/rust-lang/rust/pull/37408) | [next](https://github.com/rust-lang/rust/pull/37676)) of patches designed to rework rustc into an out-of-order on-demand pipeline model for both better feature support (e.g. [MIR-based](https://github.com/solson/miri) early constant evaluation) and incremental execution of compiler passes (e.g. type-checking), with beneficial consequences to IDE support as well.
If any motivation is unclear, please ask for additional PR description clarifications or code comments._

<hr>

The main change here is that functions and closures both use `Expr` instead of `Block` for their bodies.
For closures this actually allows a honest representation of brace-less closure bodies, e.g. `|x| x + 1` is now distinguishable from `|x| { x + 1 }`, therefore this PR is `[syntax-breaking]` (cc @Manishearth).

Using `Expr` allows more logic to be shared between constant bodies and function bodies, with some small such changes already part of this PR, and eventually easing #35078 and per-body type tables.

Incidentally, there used to be some corners cut here and there and as such I had to (re)write divergence tracking for type-checking so that it is capable of understanding basic structured control-flow:

``` rust
fn a(x: bool) -> i32 {
    // match also works (as long as all arms diverge)
    if x { panic!("true") } else { return 1; }
    0 // "unreachable expression" after this PR
}
```

And since liveness' "not all control paths return a value" moved to type-checking we can have nice things:

``` rust
// before & after:
fn b() -> i32 { 0; } // help: consider removing this semicolon

// only after this PR
fn c() -> i32 { { 0; } } // help: consider removing this semicolon
fn d() { let x: i32 = { 0; }; } // help: consider removing this semicolon
fn e() { f({ 0; }); } // help: consider removing this semicolon
```

### CODE REVIEW

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -3980,72 +4090,81 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                        }
+                        if let Some(last_stmt) = extra_semi {
+                            let original_span = original_sp(self.tcx.sess.codemap(),
+                                                            last_stmt.span, blk.span);
+                            let span_semi = Span {
+                                lo: original_span.hi - BytePos(1),
+                                hi: original_span.hi,
+                                expn_id: original_span.expn_id
+                            };
+                            err.span_help(span_semi, "consider removing this semicolon:");
```
Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: cc @GuillaumeGomez I wanted to show this on IRC but you left. This and the `report_and_explain_type_error` call above is how you can emit type errors with additional information (note/help) attached to them.

**Comment by Manishearth** [comment]
Feel free to merge this early, just ping me when you do.

File: src/librustc_typeck/check/mod.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/check/mod.rs]: Ah nice, thanks!

File: src/librustc_typeck/check/_match.rs [hunk 2]
```diff
@@ -476,11 +484,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             };
         }
 
+        // We won't diverge unless the discriminant or all arms diverge.
+        self.diverges.set(discrim_diverges | all_pats_diverge | all_bodies_diverge);
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/_match.rs]: Not that I care *too much* here, but I guess plausibly you could have 50% of the patterns diverging and the other 50% of the bodies diverging, right? 

i.e., the proper test is `discrim_diverges | all_arms_diverge`, where `all_arms_diverge = pat_diverges | body_diverges`

File: src/libpanic_abort/lib.rs [hunk 3]
```diff
@@ -53,7 +53,7 @@ pub unsafe extern fn __rust_maybe_catch_panic(f: fn(*mut u8),
 // now hopefully.
 #[no_mangle]
 pub unsafe extern fn __rust_start_panic(_data: usize, _vtable: usize) -> u32 {
-    return abort();
+    abort();
```
Comment (nikomatsakis) [inline src/libpanic_abort/lib.rs]: This is to silence a warning? Do we need to be concerned about back-compat? I guess this just causes new *warnings*, right?

**Review by nikomatsakis — APPROVED** [review]
r=me, presuming that there aren't any backcompat concerns here.

**Comment by nikomatsakis** [comment]
hmm, so I'm not sure how precise the old code was around matches -- is it worth doing a crater run or anything? Probably -- if you are with me -- you should just fix the match imprecision anyhow.

File: src/libpanic_abort/lib.rs (hunk 3 above)
Comment (eddyb) [inline src/libpanic_abort/lib.rs]: New lint warnings, which only cause problems because we use `#![deny(warnings)]`.

**Comment by eddyb** [comment]
[Crater report](https://gist.github.com/eddyb/57882da2b7027d9601b80d107ca2b6e7) consists mostly of unrelated build failures, compiler plugin fallout and [one new warning](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2FgOwn7JPOTcuXVj9rKHAS8Q%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=110&highlightEnd=110&lineNumber=&wrapLines=false&showLineNumbers=true&jumpToHighlight=true&followLog=false).

@bors r=nikomatsakis

**Comment by alexcrichton** [comment]
@bors: retry

* https://github.com/rust-lang/rust/pull/37651

**Pull request merged**